### PR TITLE
[WIP] Add Benchmarking Framework.

### LINKED ISSUE #1365: Envoy Gateway performance at scale
*Description*:
This issue tracks the performance (Throughput, Latency) of Envoy Gateway control plane and data plane at scale (Service, xRoutes, Gateway, Client Connections) . 
  * The output of this issue should be a document that can be presented to the user to better understand what is the performance of Envoy Gateway by default and how can this be increased if there is an increase in scale.
  * Other issues might need to be created (based on testing) and linked to this issue 

[optional *Relevant Links*:]
Emissary: https://www.getambassador.io/docs/emissary/latest/topics/running/scaling
Contour: https://github.com/projectcontour/contour-perf / https://projectcontour.io/guides/resource-limits/
Istio: https://istio.io/v1.16/docs/ops/deployment/performance-and-scalability/

### PR DESCRIPTION
**Pull request by gyohuangxin**
**What type of PR is this?**

**What this PR does / why we need it**:

This PR is to add benchmarking framework as discussed in [design](https://docs.google.com/document/d/1wlV81FqVVl1Gq2O-faGiiQfr-TEVPGv9zlPu0j1f000/). 

**Which issue(s) this PR fixes**:

Fixes #1365 #2325

### CODE REVIEW

File: .github/workflows/benchmark.yaml [hunk 1]
```diff
@@ -0,0 +1,54 @@
+name: Benchmarking Tests at Scale
+on:
+  pull_request:
+    branches:
+    - "main"
+    - "release/v*"
+
+jobs:
```
Comment (arkodg) [inline .github/workflows/benchmark.yaml:8]: also check out https://github.com/envoyproxy/gateway/blob/d44f4be1d60b778b8acefc2b67008a8947fa62a5/.github/workflows/build_and_test.yaml#L128

**Comment by zirain** [comment]
@gyohuangxin will you be able to finish this before KubeCon EU, will be glad to see the first report before it.

**Comment by gyohuangxin** [comment]
@zirain Currently, I add a makefile target `benchmark` for the benchmark tests. With running `make benchmark`, it will:
- Create a cluster
- Install the Envoy gateway
- Deploy the Nighthawk test server
- Apply Gateway API configurations (gatewayclass, gateway, httproute)
- Run the benchmark tests via Nighthawk load generator

And above steps have been implemented in github action runner, and these days I'm making it configurable in github action page with some configurations contain RPS, connections, test duration, httproute numbers. 

It's the first step which can be done before KubeCon EU. The next step is to use Prometheus to collect metrics from both control-plane and data-plane, then store the data somewhere and analyze them in some ways, which is risky before KubeCon EU.

What do you think about presenting the first step at KubeCon, I'm pleased prepare anything you need for the presentation, such as the diagram, demo……

**Comment by gyohuangxin** [comment]
@arkodg Thanks for explicitly dividing it to three phases, it makes sense to me. I'd like to get the Phase 1 completed soon.

**Comment by arkodg** [comment]
@gyohuangxin still working on this one ? would be great to get this into v1.1

**Comment by gyohuangxin** [comment]
@arkodg I'm sorry, I may still be busy with commanies matters in the second quarter.

**Comment by shawnh2** [comment]
> @arkodg I'm sorry, I may still be busy with commanies matters in the second quarter.

Thanks for all the wonderful works you've done, I will take this forward, feel free to reach out or take a review 😄

**Pull request closed without merging**